### DISCUSSION TITLE
adjust pwm in function of RPM sensor?

### DISCUSSION THREAD

**Discussion by elsolo5000-2** [fra589/grbl-Mega-5X#270] [Ideas]
Knowing the firmware, do you think it IS possible to adjust pwm output in function of RPM sensor?
I Saw a motor controlled and just thinked to that

**Comment by fra589** [comment]
Hi @elsolo5000-2,  
  
Can you tell me more about the RPM sensor ?  
Which kind of output it have ?  
  
@++;  
Gauthier.

**Reply by elsolo5000-2** [reply]
Hi
Some small motors have hall sensors, some people add optical sensor, it output hi-low voltage. And be able to dynamically adjust the pwm in function of workload will greatly improove engine torque at low RPM. And will give stable RPM. Some advanced systems use Line noise or current surges to determine engine RPM too but it IS an other subject.
So if the méga 2560 with grbl Can calculate spindle speed in function of a delay between hi-low signal Or take the voltage on a resistor and capacitif with an analog input without impacting the rest i thing i IS a good idea to try it😁

I Saw your name IS Gauthier, tu parle français or i must continue to make hugly phrases because i didnt liste my english teacher at School🤷

And finally i forgot, my poor man's spindle IS a 775 motor, it has no sensor but the thing passed in my head simply because it blocked Yesterday on an aluminium gasket near the final pass😱
And searching a better motor controller i thinked "but... An microcontroller can Do this!"

**Comment by fra589** [comment]
Bonjour,  
  
Ça ne me dérange pas, bien au contraire de discuter en Français, ceux qui ne le parlent pas pouront toujours utiliser l'un des multiples traducteurs disponibles de-ci de-là sur Internet. :smile:  
  
L'idée pourrait paraitre bonne si les ressources du microcontrôleur ATmega2560 étaient moins proche de leurs limites.  
En effet, pour pouvoir mesurer le temps entre les impulsions qui permettraient de calculer la vitesse de rotation de la broche à l'aide de capteurs optiques ou effet hall, il faut disposer d'un timer libre. Or, tous les timers du microcontrôleur sont déjà utilisés pour les différentes fonctions de Grbl.  
De même, le calcul de la vitesse de broche à partir d'une information analogique ou du bruit sur la ligne d'alimentation demanderait un temps de calcul trop important et aurait un impact vraiment trop négatif sur les performances et la capacité de Grbl.  
  
Le mieux pour cela est, à mon avis, d'ajouter un autre microcontrôleur intermédiaire entre Grbl et le moteur de broche dont le rôle serait d'assurer cette régulation de vitesse sans avoir à modifier Grbl.  
  
@++;  
Gauthier.

**Reply by elsolo5000-2** [reply]
humm quelle tristesse 😭 quand j ai une autre idée je reviens par ici.. et si on le faisait a partir d'un simple i/o qui dit ca tourne bien/ca ca tourne vraiment pas bien ??
